### DISCUSSION TITLE
General Advice Regarding Modules/Systems

### DISCUSSION THREAD

**Discussion by RandomEC** [evennia/evennia#2639] [General Discussion]
As I've been working through building the systems that I want for my MUD, I've given at least some degree of thought to trying to keep things orderly a) to keep myself from going insane looking for them later, and b) in case someone else ever asks for help on a subject, and I can just give them my whole code for solving that problem. While creating anything with a class (commands, objects, scripts), I've not had any issue putting that in its own module, or in a module with things related to the same problem. Let's take door commands as a good example of the issues I've had on this. For door commands, here's what I implemented:

1. A set of commands for dealing with doors - open, close, lock, unlock.
2. A set of locks for whether the above commands could be used, as well as to prevent players from going through closed doors.
3. Two dictionaries on the exit object that 1) track the current open/closed, closeable/not closeable, locked/unlocked status of the door in that exit, and 2) store the "default" status for the forgoing alternatives for when the door resets.

Item 1 was easy enough to build in my door command module. 

Item 2 I know should be possible to build in my door command module, and I tried it this way at one point, and just couldn't get it to work, even with settings referring to the location. I suspect this is just a lack of diligence/time on my part, as I didn't work at it too hard since it worked fine just defining these in the locks file.

Item 3 is my real question. I have made these changes on my standard exit object, leaving a chunk of the "project" that is not implemented through the module, but elsewhere. I can certainly explain what was done with comments in the module, even including the specific code that was used, but I can't figure out how to implement those changes in the module. Is this even possible? How do people handle this part of the project? Do you just implement in the files of objects you are impacting? Obviously this isn't something stopping me from moving forward, I'm just wondering what best practices are/everyone's opinion.

Thanks!

**Comment by InspectorCaracal** [comment]
Usually what I do for situations like 3 is instead of e.g. changing the standard Exit class, I would make a `DoorExit` class in my door module that inherits from `DefaultExit`, implement the changes there, and then change the inheritance of the base class of `Exit` to inherit from `DoorExit`. That way everything stays tidily in your door module!

As for item 2, I poked at your repository to see what you did and I would probably have used state attributes on the Exit objects instead of custom access locks. That way everything can go directly into your module.

Since I find it easier to show than explain, it'd be structured something like this:

```
class DoorExit(DefaultExit):
    def at_object_creation(self):
        self.db.open = True

    def at_traverse(self, obj, target_loc):
        if self.db.open:
            super().at_traverse(obj,target_loc)
        else:
            self.at_failed_traverse(obj)
```
